packages/perseus/src/util/snowman-utils.ts

```diff
@@ -0,0 +1,157 @@
+    WidgetType.NumericInput,
+    WidgetType.InputNumber,
+    WidgetType.Radio,
+    WidgetType.InteractiveGraph,
+    WidgetType.InteractiveNumberLine,
+    WidgetType.Categorizer,
+    WidgetType.Plotter,
+    WidgetType.Orderer,
+    WidgetType.Protractor,
+    WidgetType.Matcher,
+    WidgetType.Sorter,
+];
+
+/**
+ * Regex for widget placeholders in a string.
+ *
+ * First capture group is the widget ID (ex. 'radio 1')
+ * Second capture group is the widget type (ex. "radio)
+ * exec return will look like: ['[[☃ radio 1]]', 'radio 1', 'radio']
+ */
+export const widgetRegex = /\[\[☃ (([a-z-]+) \d+)\]\]/g;
+
+/**
+ * Add a widget placeholder using the provided widget type and instance number.
+ * ex. addWidget(WidgetType.Radio, 1) => "[[☃ radio 1]]"
+ *
+ * @param {WidgetType} widgetType
+ * @param {number} instance
+ * @returns {string}
+ */
+export function addWidget(widgetType: WidgetType, instance: number): string {
+    return `[[☃ ${widgetType} ${String(instance)}]]`;
+}
+
+/**
+ * Extract all widget IDs, which includes the widget type and instance number.
+ * example output: ['radio 1', 'categorizer 1', 'categorizor 2']
+ *
+ * Content should contain Perseus widget placeholders,
+ * which look like: '[[☃ radio 1]]'.
+ *
+ * @param {string} content
+ * @returns {Array<WidgetId>} widgets
+ */
+export function getAllWidgetIds(content: string): Array<WidgetId> {
+    const widgets: Array<WidgetId> = [];
+
+    let match = widgetRegex.exec(content);
+
+    while (match !== null) {
+        widgets.push(match[1] as WidgetId);
+        match = widgetRegex.exec(content);
+    }
+
+    return widgets;
+}
+
+function isWidgetType(value: string): value is WidgetType {
+    return Object.values(WidgetType).some((widgetType) => value === widgetType);
+}
+
+/**
+ * Extract all widget types from a Perseus JSON content string.
+ * This does not include the instance number and prevents duplicates.
+ * example output: ['radio', 'categorizer']
+ *
+ * @param {string} content
+ * @returns {Array<WidgetId>}
+ */
+export function getAllWidgetTypes(content: string): Array<WidgetId> {
+    const widgetTypes: Array<string> = [];
+
+    let match = widgetRegex.exec(content);
+
+    while (match !== null) {
+        const matchType = match[2];
+        if (!widgetTypes.includes(matchType) && isWidgetType(matchType)) {
+            widgetTypes.push(matchType);
+        }
+        match = widgetRegex.exec(content);
+    }
+    return widgetTypes as Array<WidgetId>;
+}
+
+/**
+ * Check if a specific widget is a question widget type using its ID.
+ * The widget ID includes the widget type and the instance number.
+ *
+ * @param {WidgetId} widgetId
+ * @returns {boolean}
+ */
+function isQuestionWidgetType(widgetId: WidgetId): boolean {
+    const widgetIdString = widgetId as string;
+    return (
+        QUESTION_WIDGETS.includes(widgetIdString.split(" ")[0] as WidgetType) ||
+        QUESTION_WIDGETS.includes(widgetId as WidgetType)
+    );
+}
```

Comment:
Take a look at this code as a way to maybe prevent the need for all these `as` casting things. It's still not very correct though...I'm sure there's a better way.

``` TS
const RADIO = "radio"
const PASSAGE = "passage"

const WIDGET_TYPES = ["radio", "passage"] as const;
type WidgetType = (typeof WIDGET_TYPES)[number];

const QUESTION_WIDGETS = ["radio"] as const;
type QuestionWidgetType = (typeof QUESTION_WIDGETS)[number];

type WidgetId = `${WidgetType} ${number}`;

const radioId: WidgetId = "radio 4";
const passageId: WidgetId = "passage 2";
// @ts-expect-error
const wrongId: WidgetId = "radio"; // catches when we're missing a number
// @ts-expect-error
const wrongId2: WidgetId = "rabio 4"; // catches when we misspell radio

const radioType: WidgetType = "radio";
const passageType: WidgetType = "passage";
// @ts-expect-error
const wrongType: WidgetType = "radio 4" // catches when we add a number
// @ts-expect-error
const wrongType2: WidgetId = "rabio"; // catches when we misspell radio

const radioQuestion: QuestionWidgetType = "radio";
// @ts-expect-error
const passageQuestion: QuestionWidgetType = "passage"; // catches when we think passage is a question type

function isQuestionWidgetType(widgetId: WidgetType | WidgetId): boolean {
    return (
        // Thanks to typing the params right, we can feel safe that this will work
        // either it's already a widget type ("radio".split(" ") is ["radio"])
        // or it's a valid id for a widget ("radio 1". split(" ") is ["radio", "1"])
        // so this plus the fact that it's incredibly short-lived makes it a pretty safe "any"
        QUESTION_WIDGETS.includes((widgetId.split(" ")[0] as any))
    );
}

console.log(isQuestionWidgetType(radioId));
console.log(isQuestionWidgetType(passageId));
console.log(isQuestionWidgetType(radioType));
console.log(isQuestionWidgetType(passageType));

// @ts-expect-error
console.log(isQuestionWidgetType("hello world")); // catches arbitrary strings
```

[Playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBASgQQCIEkDyMC8MBEAnAQwBMBLEHAKFElgAUEBlBhAcQFEtcAHAiCAgOYBTSlXDQYAdRRJ2AFQD6cgJq02DTgG18xMjgA03Xv2E4AujF4xq0ANwUoATy5CpJIsKhznr7AAonFxAAMykZeSVVdQBKTTAAVwBbACMhPDN7MRoYAEUAVXU5dAA5BWlZNjkNbG1CUnILKxsoe0DXHPihaDIwSXdPbxdOAJ8Q3IKGIrRS8vkGWISUtIyKBx83DyEoFCJOAAMAEgBvPs2vHwBfGGPF1LwLvczmmDqyHYAuDc8dzh16mAALDh7M8eHxBEIPl8tj9sDgwSZXAAmYEUAD0aJgAAEoBAALRCAAeLmAUAJeDwIDwWQkAHdKWABFDTt9dnDXuRbDAMdYCFBgAALLowWlCsAioQAcjwrkSJD4JEZlhgtzS6MxOPxRJJZLSlOpz3p4CZRCRnxZMLZuEIyTIgOB3MxwD5guFoqE4tpsvlEBcABs-S9dCBVs8OYMhOb+lsI78OajQcYIRGo2dY3CERDUTzNQTiUJSeT9TTYEbGSnoechuzg-bHbz+UKICKxRLLERdgQVUk7ursbi8zqi1SSyKGQII2bK7DrQRbZz687G27W16YHK+P7AxzQ+JYByOl0oD1PofuuALVXfLP6tmNQPtQXdRSR4nwcIz8fwKfOufetGr1+TNTC5Hkl1dZt3U9VwoAFRUAGsYGA1x5WVABHX8v3FNpVmCeIwFJHoYHlT8ekvCM-FpADmQA2MAB9pyIaJPmSEAQD9IQCHFI4KBgPiXi2eI8HFPxeP48SeTkAUuPg5soBAGBAkVARFKFJCCEIRJmzwEgBAFKBDDXZ1xWCIQhEDfhTNUvlVNQqiAxFKl4LE8S+J5IQSFgtJiKgSVmwIP0ZWIRxlSos5FPWPw-j0AA6X0-U8qLcGiYjm1qYNzGiFzXJ5KkfL85UADcAvcYjdmCPKuzCzwYCijkYAARhwGKYHixKcGS1KYHS29DBwJqzCy1z+J5CAFNg1CuD9eI5LU4ICFJazYE8grFWAGVSGSP0QogAUqTJBLCqEXZEgIeDhU85UuBlKAnFaggrJwLjHEoYa+PyQoSjKcJKgYGK1umoguj8SjqKIOKpva5LNAABkafywEcaIhv46J7AuXdIHYoQYr9EABD8EjMLI2ifD8DkdmRkFxGx3H8cJiBSIvUmXD8ZDKbRmlabxgmiaPEm0zJ8MfCprmOLp3nGeJ5nBdZ5CI1Fvtc0fQs9VfGnxZ5hmmf-WWhCioUAwU2kqT9IgcCpxcXSbSw8FtKBCDwHaHeUiAgA)

Reply:
Thanks for writing out this example! Is this what it could partially look like if I used the values found in the enum utils file? Those aren't set as consts though. Would I need to create `const WIDGET_TYPES = ["radio", "passage"] as const;`with all the widgets listed and `const QUESTION_WIDGETS = ["radio"] as const;` with all the question type widgets listed? Also, if I am still declaring the type in this version, isn't it the same? Or is just having as few as possible still better?

Reply:
Will move forward with the current version and will try to remove the type declarations at a later time.